src/D2L.CodeStyle.Analyzers/Immutability/ImmutableTypeInfo.cs

```diff
@@ -0,0 +1,35 @@
+﻿using System;
+using Microsoft.CodeAnalysis;
+
+namespace D2L.CodeStyle.Analyzers.Immutability {
+	/// <summary>
+	/// A representation of the immutability of some type.
+	/// </summary>
+	internal readonly struct ImmutableTypeInfo : IEquatable<ImmutableTypeInfo> {
+		public ImmutableTypeInfo(
+			ImmutableTypeKind kind,
+			ITypeSymbol type

```

Comment:
To support generic container classes like `ImmtuableArray<T>` (either "blessed" or user-defined) this needs to have a list of `ITypeParameterSymbol`s that the immutability is conditional upon.

There won't be support for saying `X<A,B>` is `ImmutableTypeKind.Total` if `A`&`B` are `ImmutableKind.Total`, but `ImmutableTypeKind.Instance` if only `A` is, etc.... in other words we're only going to bother having at most one `ImmutableTypeInfo` judgement per `ITypeSymbol`.